src/net.cpp

```diff
@@ -1811,12 +1811,76 @@ void ThreadOpenConnections()
 
     // Initiate network connections
     int64_t nStart = GetTime();
-    while (true) {
+    unsigned int nDisconnects = 0;
+    while (true)
+    {
         ProcessOneShot();
 
         MilliSleep(500);
 
-        CSemaphoreGrant grant(*semOutbound);
+        // Only connect out to one peer per network group (/16 for IPv4).
+        // Do this here so we don't have to critsect vNodes inside mapAddresses critsect.
+        // And also must do this before the semaphore grant so that we don't have to block
+        // if the grants are all taken and we want to disconnect a node in the event that
+        // we don't have enough connections to XTHIN capable nodes yet.
+        int nOutbound = 0;
+        int nThinBlockCapable = 0;
+        set<vector<unsigned char> > setConnected;
+        CNode* ptemp = nullptr;
+        bool fDisconnected = false;
+        {
+            LOCK(cs_vNodes);
+            BOOST_FOREACH (CNode* pnode, vNodes)
+            {
+                if (pnode->fOutbound) // only count outgoing connections.
```

Comment:
> I think it make the network stronger to have 4 outgoing because then we'll pretty much have 4 outs and 4 ins as everyone upgrades. But other than that, what should the best count be?

Not sure if I made my suggestion clear as your answer doesn't seem related.

Your PR is about the node having a connection to a certain number of xthin nodes. This is great.
Your code, however, doesn't count incoming connections to get support. And I don't see why you would continue disconnecting nodes if you already have your allotted number of xthin peers, but they are incoming.

It would not make any difference to the behaviour, you still reach the right number of xthin peers. It doesn't matter if an incoming or an ourgoing peer disconnected, you would behave the same to get the number of xthin peers up again.